File: 01052. Grumpy Bookstore Owner/solution.js
/**
 * Problem: 1052. Grumpy Bookstore Owner
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Calculates the maximum satisfied customers using the technique
 *
 * @param {number[]} customers - Customers per minute
 * @param {number[]} grumpy - 1 if grumpy, 0 otherwise
 * @param {number} minutes - Technique duration
 *
 * @returns {number} - Maximum satisfied customers
 */
const maxSatisfied = (customers, grumpy, minutes) => {
  // Initialize the maximum additional satisfied customers using the technique
  let maxAdditionalSatisfied = 0
  // Initialize the total satisfied customers without using the technique
  let totalSatisfied = 0

  // Calculate initial window (first 'minutes') of additional satisfied customers
  for (let i = 0; i < minutes; i++) {
    // If owner is grumpy, these customers can be satisfied by the technique
    if (grumpy[i] === 1) maxAdditionalSatisfied += customers[i]
    // If owner is not grumpy, these customers are always satisfied
    else totalSatisfied += customers[i]
  }

  // Initialize left pointer for sliding window
  let left = 0
  // Current additional satisfied customers in the window
  let currentAdditionalSatisfied = maxAdditionalSatisfied

  // Slide the window through the rest of the array
  for (let right = minutes; right < customers.length; right++) {
    // Remove the effect of the leftmost minute if owner was grumpy
    if (grumpy[left] === 1) currentAdditionalSatisfied -= customers[left]

    // Move the window forward
    left++

    // Add the effect of the new rightmost minute
    if (grumpy[right] === 1) currentAdditionalSatisfied += customers[right]
    // If owner is not grumpy, these customers are always satisfied
    else totalSatisfied += customers[right]

    // Update the maximum additional satisfied customers
    maxAdditionalSatisfied = Math.max(
      maxAdditionalSatisfied,
      currentAdditionalSatisfied
    )
  }

  // Return the sum of always satisfied and maximum additional satisfied customers
  return totalSatisfied + maxAdditionalSatisfied
}

File: 01052. Grumpy Bookstore Owner/solution.go
/**
 * Problem: 1052. Grumpy Bookstore Owner
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func maxSatisfied(customers []int, grumpy []int, minutes int) int {
	// Initialize totalSatisfied to store the number of customers already satisfied (when owner is not grumpy)
	totalSatisfied := 0
	// prefixGrumpySum will store prefix sums of customers who are unsatisfied due to grumpiness
	prefixGrumpySum := make([]int, len(customers)+1)

	// Iterate through each minute
	for i := range customers {
		if grumpy[i] == 0 {
			// Owner is not grumpy, add customers to totalSatisfied
			totalSatisfied += customers[i]
		} else {
			// Owner is grumpy, add customers to prefixGrumpySum for possible satisfaction later
			prefixGrumpySum[i+1] = customers[i]
		}

		// Build prefix sum for grumpy customers
		if i > 0 {
			prefixGrumpySum[i+1] += prefixGrumpySum[i]
		}
	}

	// Initialize maxSatisfied to the base satisfied customers
	maxSatisfied := totalSatisfied

	// Try every possible window of 'minutes' to maximize satisfied customers
	for i := 0; i <= len(customers)-minutes; i++ {
		// Calculate satisfied customers if owner is not grumpy for this window
		maxSatisfied = max(maxSatisfied, totalSatisfied+prefixGrumpySum[i+minutes]-prefixGrumpySum[i])
	}

	// Return the maximum number of satisfied customers
	return maxSatisfied
}
